src/Data/Tree/DUAL/Internal.hs

```diff
@@ -146,178 +119,137 @@ pullU t@(Annot _ (DUALTreeU (u, _))) = pack (u, t)
-           -> (d -> r -> r)      -- ^ Process an internal d node
-           -> (a -> r -> r)      -- ^ Process an internal datum
-           -> DUALTreeNE d u a l -> r
-foldDUALNE  = foldDUALNE' (Option Nothing)
+-- | Fold a dual tree for a monoidal result @r@. The @d@ annotations are
+--   accumulated from the top of the tree. Static @a@ annotations are
+--   acted on by the @d@ annotation accumulated up to that point.
+foldDUAL :: (Action d a, Monoid' d, Monoid r)
+         => (d -> l -> r) -- ^ Process a leaf
+         -> (a -> r -> r) -- ^ Process an annotation
+         -> DUALTree d u a l
+         -> r
+foldDUAL _  _  EmptyDUAL       = mempty
+foldDUAL lF aF (DUALTree _ t0) = go mempty t0
   where
-    foldDUALNE' dacc lf _   _   _    _   (Leaf _ l)  = lf (option mempty id dacc) l
-    foldDUALNE' _    _  lfU _   _    _   (LeafU _)   = lfU
-    foldDUALNE' dacc lf lfU con down ann (Concat ts)
-      = con (NEL.map (foldDUALNE' dacc lf lfU con down ann . snd . unpack) ts)
-    foldDUALNE' dacc lf lfU con down ann (Act d t)
-      = down d (foldDUALNE' (dacc <> (Option (Just d))) lf lfU con down ann . snd . unpack $ t)
-    foldDUALNE' dacc lf lfU con down ann (Annot a t)
-      = ann a (foldDUALNE' dacc lf lfU con down ann . snd . unpack $ t)
-
--- | Fold for DUAL-trees. It is given access to the internal and leaf
---   data, internal @d@ values, and the accumulated @d@ values at each
---   leaf.  It is also allowed to replace \"@u@-only\" leaves with a
---   constant value.  In particular, however, it is /not/ given access
---   to any of the @u@ annotations, the idea being that those are used
---   only for /constructing/ trees.  If you do need access to @u@
---   values, you can duplicate the values you need in the internal
---   data nodes.
---
---   Be careful not to mix up the @d@ values at internal nodes with
---   the @d@ values at leaves.  Each @d@ value at a leaf satisfies the
---   property that it is the 'mconcat' of all internal @d@ values
---   along the path from the root to the leaf.
---
---   The result is @Nothing@ if and only if the tree is empty.
-foldDUAL :: (Semigroup d, Monoid d)
-         => (d -> l -> r)          -- ^ Process a leaf datum along with the
-                                   --   accumulation of @d@ values along the
-                                   --   path from the root
-         -> r                      -- ^ Replace @u@-only nodes
-         -> (NonEmpty r -> r)      -- ^ Combine results at a branch node
-         -> (d -> r -> r)          -- ^ Process an internal d node
-         -> (a -> r -> r)          -- ^ Process an internal datum
-         -> DUALTree d u a l -> Maybe r
-foldDUAL _ _ _ _ _ (DUALTree (Option Nothing))
-  = Nothing
-foldDUAL l u c d a (DUALTree (Option (Just (DUALTreeU (_, t)))))
-  = Just $ foldDUALNE l u c d a t
+    go !d = \case
+      Down d' t -> go (d <> d') t
+      Leaf l    -> lF d l
+      Annot a t -> aF (act d a) (go d t)
```

Comment:
Fair enough, I'm happy with that.  It's always easy enough to give a trivial identity action if you have some annotation type that you can't or don't want to make transformable.